lib/tree.rb

```diff
@@ -16,47 +16,168 @@ def initialize
     @root = nil
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
-  def add(key, value)
-    raise NotImplementedError
+  # Time Complexity: O(log(n)) - Assuming the tree is somewhat balanced, this could potentially be O(log(n)) because we may have to traverse the height of the tree to get to a leaf node. 
+  # Space Complexity: O(n) - for recursive solution, there is space in the form of the call stack
+  # Space Complexity: O(1) - for while loop
+  
+  def add(key, value = nil)
+    new_node = TreeNode.new(key, value)
+    if @root.nil?
+      @root = new_node
+    else
+      add_helper(@root, new_node)
+    end
+
   end
 
-  # Time Complexity: 
-  # Space Complexity: 
+  def add_helper(current, new_node)
+    return new_node if current.nil?
+
+    if current.key > new_node.key
+      current.left = add_helper(current.left, new_node)
+    else
+      current.right = add_helper(current.right, new_node)
+    end
+    return current
+
+    
+    # try it with a while loop
+    # until current.nil?
+    #   if current.value > new_node.value && current.left.nil?
+    #     current.left = new_node
+    #     return
+    #   elsif current.value > new_node.value
+    #     current = current.left
+    #   elsif current.value <= new_node.value && current.right.nil?
+    #     current.right = new_node
+    #     return
+    #   elsif current.value <= new_node.value
+    #     current = current.right
+        
+    #   end
+    # end
+    # return current
+  end
+
+  # Time Complexity: O(log(n))for balanced tree, O(n) if not.
+  # Space Complexity: O(n) for call stack. 
   def find(key)
```

Comment:
👍 Both the space/time complexities are O(n) if the tree is unbalanced and O(log n) if the tree is balanced.